swallowtail/utils/binary.py

```diff
@@ -0,0 +1,215 @@
+    
+    def invalid_reason(self) -> str | None:
+        if not self.name: return 'fields must have names'
+        if self.array_length is not None and self.array_length <= 0: return 'cannot have an array length of 0'
+        if self.size <= 0: return 'cannot have a size of 0 (make sure your fields are annotated)'
+        if len(self.format) <= 0: return 'format resolved to empty string'
+        if self.structure is not None and not issubclass(self.structure, Structure): return f'invalid field type {self.structure!r}'
+        if self.structure is not None and self.type_cast is not None and self.type_cast != self.structure: return 'conflicting type_cast and structure'
+        return None
+
+    def valid(self) -> bool:
+        return self.invalid_reason() is None
+
+@dataclasses.dataclass
+class StructureMetadata:
+    size: int = 0
+    format: str = ''
+    fields: list[StructuredFieldMetadata] = dataclasses.field(default_factory=list) # pls don't modify directly
+
+    def add_field(self, field: StructuredFieldMetadata) -> None:
+        self.fields.append(field)
+        self.size += field.size
+        self.format += field.format
+
+class BYTE(int, StructuredField):
+    size: int = 1
+    format: str = 'B'
+
+class CHAR(bytes, StructuredField):
+    size: int = 1
+    format: str = 'c'
+
+    def __repr__(self):
+        try:
+            return repr(self.decode())
+        except:
+            return super().__repr__()
+
+class STRING(bytes): # notice it does not inherit from StructuredField since it's hardcoded
+    def __repr__(self):
+        try:
+            human_readable = self.rstrip(b'\x00').decode()
+            if not human_readable or not human_readable.isprintable():
+                raise ValueError
+            else:
+                return repr(human_readable)
+        except:
+            return super().__repr__()
+
+class WORD(int, StructuredField):
+    size: int = 2
+    format: str = 'H'
+
+    def __repr__(self):
+        return f'{self:0{self.size * 2}x}h ({self:d})'
+
+class LONG(int, StructuredField):
+    size: int = 4
+    format: str = 'l'
+
+class DWORD(int, StructuredField):
+    size: int = 4
+    format: str = 'L'
+
+    def __repr__(self):
+        return f'{self:0{self.size * 2}x}h ({self:,d})'
+
+def LENGTH(length: int) -> typing.Any:
+    return length
+
+class Structure:
+    '''
+    >>> class Test(Structure):
+    ...     x: BYTE
+    ...     y: CHAR
+    ...     z: DWORD
+    ...     w: list[CHAR] = LENGTH(3)
+    ...     p: STRING = LENGTH(11)
+    ...
+    >>> Test._metadata.size
+    20
+    >>> Test._metadata.format
+    'BcL3c11s'
+    >>> Test(b'x')
+    Traceback (most recent call last):
+    ...
+    ValueError: This structure requires exactly 20 bytes, but you provided 1.
+    >>> t = Test(b'\\xFFH\\x05\\x00\\x00\\x00ABCHELLO WORLD')
+    >>> t.x, t.y, t.z, t.w, t.p
+    (255, 'H', 0x5, ['A', 'B', 'C'], 'HELLO WORLD')
```

Comment:
I might move this into a repo of its own, because that user-code is absolutely stunning.